src/bootstrap/compile.rs

```diff
@@ -311,8 +309,18 @@ pub fn std_cargo(builder: &Builder<'_>, target: TargetSelection, stage: u32, car
     // If `compiler-rt` is available ensure that the `c` feature of the
     // `compiler-builtins` crate is enabled and it's configured to learn where
     // `compiler-rt` is located.
-    let compiler_builtins_root = builder.src.join("src/llvm-project/compiler-rt");
-    let compiler_builtins_c_feature = if compiler_builtins_root.exists() {
+    let compiler_builtins_c_feature = if builder.config.optimized_compiler_builtins {
+        // NOTE: this interacts strangely with `llvm-has-rust-patches`. In that case, we enforce `submodules = false`, so this is a no-op.
+        // But, the user could still decide to manually use an in-tree submodule.
+        //
+        // Using system llvm is not supported.
+        builder.update_submodule(&Path::new("src").join("llvm-project"));
+        let compiler_builtins_root = builder.src.join("src/llvm-project/compiler-rt");
+        if !compiler_builtins_root.exists() || builder.is_system_llvm(target) {
+            panic!(
+                "needed LLVM sources available to build `compiler-rt`, but they weren't present; consider enabling `build.submodules = true`, disabling `optimized-compiler-builtins`, or unsetting `llvm-config`"
+            );
+        }
```

Comment:
Since `optimized-compiler-builtins` is required to produce a working compiler on the platforms we support, there should be a way to set it with a custom/system LLVM. Maybe having a `llvm.compiler-rt-source` setting to specify the path to the custom compiler-rt source code would work?

Reply:
This was *already* unsupported, it was just unclear because everything was implicit. The former logic was to enable the option when `llvm-project` was checked out and disable it when it's not. I don't think we should try and expand that here.

Reply:
The lack of an escape hatch prevents building the compiler on a Tier 1 platform though, while before it worked (maybe by accident). I'm happy to contribute a `llvm.compiler-rt-source` flag myself.